File: src/test/java/net/lumalyte/trivia/models/PlayerStatsTest.java
package net.lumalyte.trivia.models;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.util.UUID;

public class PlayerStatsTest {
    private PlayerStats stats;
    private final UUID playerId = UUID.randomUUID();
    private final String playerName = "TestPlayer";

    @BeforeEach
    void setUp() {
        stats = new PlayerStats(playerId, playerName);
    }

    @Test
    void testInitialValues() {
        assertEquals(0, stats.getTotalCorrect());
        assertEquals(0, stats.getEasyCorrect());
        assertEquals(0, stats.getMediumCorrect());
        assertEquals(0, stats.getHardCorrect());
        assertEquals(0, stats.getPoints());
        assertEquals(playerId, stats.getPlayerId());
        assertEquals(playerName, stats.getPlayerName());
    }

    @Test
    void testAddCorrectAnswer() {
        // Test easy question
        stats.addCorrectAnswer("easy");
        assertEquals(1, stats.getTotalCorrect());
        assertEquals(1, stats.getEasyCorrect());
        assertEquals(0, stats.getMediumCorrect());
        assertEquals(0, stats.getHardCorrect());
        assertEquals(1, stats.getPoints());

        // Test medium question
        stats.addCorrectAnswer("medium");
        assertEquals(2, stats.getTotalCorrect());
        assertEquals(1, stats.getEasyCorrect());
        assertEquals(1, stats.getMediumCorrect());
        assertEquals(0, stats.getHardCorrect());
        assertEquals(3, stats.getPoints()); // 1 + 2 points

        // Test hard question
        stats.addCorrectAnswer("hard");
        assertEquals(3, stats.getTotalCorrect());
        assertEquals(1, stats.getEasyCorrect());
        assertEquals(1, stats.getMediumCorrect());
        assertEquals(1, stats.getHardCorrect());
        assertEquals(6, stats.getPoints()); // 1 + 2 + 3 points
    }

    @Test
    void testCompareTo() {
        PlayerStats otherStats = new PlayerStats(UUID.randomUUID(), "OtherPlayer");
        
        // Both have 0 points
        assertEquals(0, stats.compareTo(otherStats));
        
        // This player has more points
        stats.addCorrectAnswer("hard"); // 3 points
        assertTrue(stats.compareTo(otherStats) < 0); // Should be "less than" for descending order
        
        // Other player has more points
        otherStats.addCorrectAnswer("hard"); // 3 points
        otherStats.addCorrectAnswer("medium"); // +2 points
        assertTrue(stats.compareTo(otherStats) > 0); // Should be "greater than" for descending order
    }
} 

File: src/test/kotlin/net/badgersmc/trivia/domain/PlayerStatsTest.kt
package net.badgersmc.trivia.domain

import org.junit.jupiter.api.Test
import java.util.UUID
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class PlayerStatsTest {
    private val playerId = UUID.randomUUID()
    private val playerName = "TestPlayer"

    @Test
    fun `initial values are zero`() {
        val stats = PlayerStats(playerId, playerName)
        assertEquals(0, stats.totalCorrect)
        assertEquals(0, stats.easyCorrect)
        assertEquals(0, stats.mediumCorrect)
        assertEquals(0, stats.hardCorrect)
        assertEquals(0, stats.points)
        assertEquals(playerId, stats.playerId)
        assertEquals(playerName, stats.playerName)
    }

    @Test
    fun `addCorrectAnswer easy adds 1 point`() {
        val stats = PlayerStats(playerId, playerName)
        stats.addCorrectAnswer("easy")
        assertEquals(1, stats.totalCorrect)
        assertEquals(1, stats.easyCorrect)
        assertEquals(0, stats.mediumCorrect)
        assertEquals(0, stats.hardCorrect)
        assertEquals(1, stats.points)
    }

    @Test
    fun `addCorrectAnswer medium adds 2 points`() {
        val stats = PlayerStats(playerId, playerName)
        stats.addCorrectAnswer("medium")
        assertEquals(1, stats.totalCorrect)
        assertEquals(0, stats.easyCorrect)
        assertEquals(1, stats.mediumCorrect)
        assertEquals(0, stats.hardCorrect)
        assertEquals(2, stats.points)
    }

    @Test
    fun `addCorrectAnswer hard adds 3 points`() {
        val stats = PlayerStats(playerId, playerName)
        stats.addCorrectAnswer("hard")
        assertEquals(1, stats.totalCorrect)
        assertEquals(3, stats.points)
    }

    @Test
    fun `multiple answers accumulate correctly`() {
        val stats = PlayerStats(playerId, playerName)
        stats.addCorrectAnswer("easy")
        stats.addCorrectAnswer("medium")
        stats.addCorrectAnswer("hard")
        assertEquals(3, stats.totalCorrect)
        assertEquals(1, stats.easyCorrect)
        assertEquals(1, stats.mediumCorrect)
        assertEquals(1, stats.hardCorrect)
        assertEquals(6, stats.points) // 1 + 2 + 3
    }

    @Test
    fun `compareTo sorts by points descending`() {
        val a = PlayerStats(UUID.randomUUID(), "A")
        val b = PlayerStats(UUID.randomUUID(), "B")

        // Equal points → tie
        assertEquals(0, a.compareTo(b))

        // A has more points → A < B (descending)
        a.addCorrectAnswer("hard") // 3 points
        assertTrue(a.compareTo(b) < 0)

        // B has more points → A > B (descending)
        b.addCorrectAnswer("hard") // 3
        b.addCorrectAnswer("medium") // +2 = 5
        assertTrue(a.compareTo(b) > 0)
    }

    @Test
    fun `playerName is preserved`() {
        val stats = PlayerStats(playerId, "Alice")
        assertEquals("Alice", stats.playerName)
        stats.addCorrectAnswer("easy")
        assertEquals("Alice", stats.playerName)
    }
}
